bionic/aip/main.py

```diff
@@ -0,0 +1,70 @@
+"""
+This module is run as main in order to execute a task on a worker.
+"""
+
+import logging
+import os
+import sys
+
+from bionic.deps.optdep import import_optional_dependency
+
+
+def run():
+    # Scope the import to this function to avoid raising for anyone not using it.
+    blocks = import_optional_dependency("blocks", raise_on_missing=True)
+    cloudpickle = import_optional_dependency("cloudpickle", raise_on_missing=True)
+
+    ipath = sys.argv[-1]
+    fs = blocks.filesystem.GCSNativeFileSystem()
+    with fs.open(ipath, "rb") as f:
+        task = cloudpickle.load(f)
+
+    # Now that we have the task, setup logging.
+    _setup_logging(task.job_id(), task.config.project)
+    logging.info(f"Read task from {ipath}")
+
+    logging.info("Starting execution")
+    result = task.function()
+
+    opath = task.output_uri()
+    logging.info(f"Uploading result to {opath}")
+    blocks.pickle(result, opath)
+    return
+
+
+def _setup_logging(job_id, project):
+    # TODO This is the ID of the hyperparameter tuning trial currently
+    # running on this VM. This field is only set if the current
+    # training job is a hyperparameter tuning job. Conductor uses this
+    # environment variable but AIP documentation suggests us to use
+    # TF_CONFIG. Check whether we need to update this env variable.
+    # Find more details on TF_CONFIG at this link:
+    # https://cloud.google.com/ai-platform/training/docs/distributed-training-details
+    trial = os.environ.get("CLOUD_ML_TRIAL_ID", None)
+
+    glogging = import_optional_dependency("google.cloud.logging", raise_on_missing=True)
+
+    client = glogging.Client(project=project)
+    resource = glogging.resource.Resource(
+        type="ml_job",
+        labels=dict(job_id=job_id, project_id=project, task_name="master-replica-0"),
+    )
+    labels = None
+    if trial is not None:
+        # Enable the grouping by trial when present.
+        labels = {"ml.googleapis.com/trial_id": trial}
+
+    # Enable only the cloud logger to avoid duplicate messages.
+    handler = glogging.handlers.handlers.CloudLoggingHandler(
+        client, resource=resource, labels=labels
+    )
+    root_logger = logging.getLogger()
+    root_logger.handlers = []  # remove stream handler
+    root_logger.setLevel(logging.DEBUG)
+    root_logger.addHandler(handler)
+    for logger_name in glogging.handlers.handlers.EXCLUDED_LOGGER_DEFAULTS:
```

Comment:
What does this do?

Reply:
This excludes loggers that the library uses itself to avoid infinite recursion. We do the same thing here as what the library provided [setup_logging](https://googleapis.dev/python/logging/1.14.0/_modules/google/cloud/logging/handlers/handlers.html) function does, but we also remove the StreamHandler as well to avoid those error logs (I added a comment in the code for this). Hence, we have this implementation here instead of using their provided method.